SU2_CFD/include/numerics_simd/CNumericsSIMD.hpp

```diff
@@ -0,0 +1,594 @@
+                                           const Gradient_t& gradient) {
+  auto vars = gatherVariables<nVar>(iPosize_t, field);
+  for (size_t iVar = 0; iVar < nVar; ++iVar) {
+    vars(iVar) += dot(gradient.innerIter<nVar,nDim>(iPosize_t,iVar), vector_ij);
+  }
+  return vars;
+}
+
+template<size_t nVar, size_t nDim, class Field_t, class Gradient_t>
+FORCEINLINE VectorDbl<nVar> musclPosize_tLimited(Int iPosize_t,
+                                              const VectorDbl<nDim>& vector_ij,
+                                              const Field_t& field,
+                                              const Field_t& limiter,
+                                              const Gradient_t& gradient) {
+  auto vars = gatherVariables<nVar>(iPosize_t, field);
+  auto itLim = limiter.innerIter<nVar>(iPosize_t);
+  for (size_t iVar = 0; iVar < nVar; ++iVar) {
+    vars(iVar) += *(itLim++) * dot(gradient.innerIter<nVar,nDim>(iPosize_t,iVar), vector_ij);
+  }
+  return vars;
+}
+
+template<size_t nDim, size_t nVar, class Field_t, class Gradient_t>
+FORCEINLINE void musclEdgeLimited(Int iPosize_t,
+                                  Int jPosize_t,
+                                  const VectorDbl<nDim>& vector_ij,
+                                  const Field_t& field,
+                                  const Gradient_t& gradient,
+                                  VectorDbl<nVar>& vars_i,
+                                  VectorDbl<nVar>& vars_j) {
+  vars_i = gatherVariables<nVar>(iPosize_t, field);
+  vars_j = gatherVariables<nVar>(jPosize_t, field);
+
+  for (size_t iVar = 0; iVar < nVar; ++iVar) {
+    auto proj_i = dot(gradient.innerIter<nVar,nDim>(iPosize_t,iVar), vector_ij);
+    auto proj_j = dot(gradient.innerIter<nVar,nDim>(jPosize_t,iVar), vector_ij);
+    auto delta_ij = vars_j(iVar) - vars_i(iVar);
+    auto delta_ij_2 = pow(delta_ij,2);
+    /// TODO: Customize the limiter function
+    auto lim_i = (delta_ij_2 + 2*proj_i*delta_ij) / (4*pow(proj_i,2) + delta_ij_2 + eps);
+    auto lim_j = (delta_ij_2 + 2*proj_j*delta_ij) / (4*pow(proj_j,2) + delta_ij_2 + eps);
+    vars_i(iVar) += lim_i * proj_i;
+    vars_j(iVar) -= lim_j * proj_j;
+  }
+}
+
+template<class T>
+struct CPair {
+  T i, j;
+};
+
+template<size_t nDim>
+struct CCompressiblePrimitives {
+  enum : size_t {nVar = nDim+4};
+  VectorDbl<nVar> all;
+  Double& temperature = all(0);
+  Double& pressure = all(nDim+1);
+  Double& density = all(nDim+2);
+  Double& enthalpy = all(nDim+3);
+  FORCEINLINE Double& velocity(size_t iDim) { return all(iDim+1); }
+  FORCEINLINE Double* velocity() { return &velocity(0); }
+};
+
+template<size_t nDim>
+struct CCompressibleConservatives {
+  enum : size_t {nVar = nDim+2};
+  VectorDbl<nVar> all;
+  Double& density = all(0);
+  Double& rhoEnergy = all(nDim+1);
+  FORCEINLINE Double& momentum(size_t iDim) { return all(iDim+1); }
+  FORCEINLINE Double* momentum() { return &momentum(0); }
+};
```

Comment:
I'm also trying to address the magic indexes (iDim+1) by defining specific types for primitive and conservative variables rather that using raw arrays.